File: ios/bridge/ExpoBridgeLibraryMutations.swift
import Foundation

@available(iOS 16.0, *)
enum ExpoBridgeLibraryMutations {
  static func addToLibrary(
    service: LibraryMutationsService,
    musicUserToken: String,
    resourceIds: [String: [String]]
  ) async throws {
    try await AppleMusicBridgeError.rethrow {
      try await service.addToLibrary(musicUserToken: musicUserToken, resourceIds: resourceIds)
    }
  }

  static func createLibraryPlaylist(
    service: LibraryMutationsService,
    musicUserToken: String,
    options: [String: Any]
  ) async throws -> [String: Any] {
    try await AppleMusicBridgeError.rethrow {
      let name = options["name"] as? String ?? ""
      let description = options["description"] as? String
      let isPublic = options["isPublic"] as? Bool ?? false
      let tracks = options["tracks"] as? [[String: String]]
      return try await service.createPlaylist(
        musicUserToken: musicUserToken,
        name: name,
        description: description,
        isPublic: isPublic,
        tracks: tracks
      )
    }
  }

  static func addTracksToLibraryPlaylist(
    service: LibraryMutationsService,
    musicUserToken: String,
    playlistId: String,
    tracks: [[String: String]]
  ) async throws {
    try await AppleMusicBridgeError.rethrow {
      try await service.addTracksToPlaylist(
        musicUserToken: musicUserToken,
        playlistId: playlistId,
        tracks: tracks
      )
    }
  }
}

File: android/src/main/java/expo/modules/applemusic/bridge/ExpoBridgeLibraryMutations.kt
package expo.modules.applemusic.bridge

import expo.modules.applemusic.LibraryMutationsRestClient
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinitionBuilder

internal fun ModuleDefinitionBuilder.registerLibraryMutationsBridge(
  libraryMutations: () -> LibraryMutationsRestClient,
) {
  AsyncFunction("addToLibrary") Coroutine { musicUserToken: String, resourceIds: Map<String, List<String>> ->
    libraryMutations().addToLibrary(musicUserToken, resourceIds)
  }

  AsyncFunction("createLibraryPlaylist") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val name = options["name"] as? String ?: ""
    val description = options["description"] as? String
    val isPublic = options["isPublic"] as? Boolean ?: false
    @Suppress("UNCHECKED_CAST")
    val tracks = options["tracks"] as? List<Map<String, String>>
    libraryMutations().createLibraryPlaylist(musicUserToken, name, description, isPublic, tracks)
  }

  AsyncFunction("addTracksToLibraryPlaylist") Coroutine {
      musicUserToken: String,
      playlistId: String,
      tracks: List<Map<String, String>>,
    ->
    libraryMutations().addTracksToLibraryPlaylist(musicUserToken, playlistId, tracks)
  }
}
